File: src/FuseProtector.cc
// 有FuseProtector类和ServiceState类
#include "./include/fuse/FuseProtector.h"

//////////////////////////////// ServiceState ////////////////////////////////////
void ServiceState::incrRequest() {
    request.fetch_add(2); // 原子变量的增加1
    excepts.fetch_add(1); // +todo 为了计算？
    // +todo:我没看明白为什么请求和异常都+1，所以我就把源代码改了。
    int requestNum = request.load();
    int exceptsNum = excepts.load();
    // 熔断比率K默认为1.2
    if (this->fuseState != CLOSE) {
        float rate = (requestNum - k*(requestNum - exceptsNum)) / (requestNum + 1);
        this->interceptRate = rate; 
        if (rate < 0.1)
            this->fuseState = CLOSE; // 熔断关闭
    }
}

void ServiceState::incrExcepts() {
    request.fetch_add(1);
    excepts.fetch_add(1);
    int requestNum = request.load();
    int exceptsNum = excepts.load();
    float rate = (requestNum - k*(requestNum - exceptsNum)) / (requestNum + 1);
    this->interceptRate = rate;
    if (rate < 0.1)
        this->fuseState = CLOSE;
    else if (rate > 0.3) {
        std::cout << "熔断器当前切换至半开" << std::endl;
        this->fuseState = HALF_OPEN;
    } else if (rate > 0.7) {
        std::cout << "熔断器当前切换至全开" << std::endl;
        this->fuseState = FALL_OPEN;
    }
}

//////////////////////////////// FuseProtector ////////////////////////////////////
// 服务名-对应一堆服务节点
// 这个是初始化，下面那个是刷新
void FuseProtector::initCache(std::unordered_map<std::string, std::set<ServiceAddress>> serviceList) {
    std::lock_guard<std::mutex> lock(mtx);
    for (auto& entry : serviceList) {
        std::string serviceName = entry.first;
        ServiceState serviceState(serviceName);
        serviceStateCache[serviceName] = serviceState;
    }
}

bool FuseProtector::fuseHandle(std::string serviceName) { // 检查当前熔断状态
    std::lock_guard<std::mutex> lock(mtx);
    ServiceState serviceState = serviceStateCache[serviceName];
    switch (serviceState.getFuseState()) {
        case CLOSE: return true;
        case FALL_OPEN: return false;
        case HALF_OPEN:
            return (rand() / (RAND_MAX + 1.0)) > serviceState.getInterceptRate();
    }
    return true;
}

// 添加成功请求
void FuseProtector::incrSuccess(std::string serviceName) {
    std::lock_guard<std::mutex> lock(mtx); // 上锁
    ServiceState& serviceState = serviceStateCache[serviceName];
    serviceState.incrRequest(); // 会+2总请求，+1异常请求
}

// 添加异常请求
void FuseProtector::incrExcept(std::string serviceName) {
    std::lock_guard<std::mutex> lock(mtx); // 上锁
    ServiceState& serviceState = serviceStateCache[serviceName];
    serviceState.incrExcepts(); // 会+1总请求，+1异常请求
}

// 数据定期刷新，确保全开状态下进行重新检查
//+todo 这里没有实现定期刷新，只是调用时刷新，可以在别的调用它的地方实现定期刷新
void FuseProtector::refreshCache() {
    std::lock_guard<std::mutex> lock(mtx); // 上锁
    for (auto& entry : serviceStateCache) {
        entry.second = ServiceState(entry.first);
    }
}

File: 项目讲解/easy_rpc/fuse/FuseProtector.java
package com.rpc.easy_rpc_govern.fuse;


import com.rpc.domain.pojo.ServiceMeta; // ServiceMeta
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Async; // 异步执行，不会阻塞当前线程。
import org.springframework.scheduling.annotation.Scheduled; // 定时数据同步。
import org.springframework.stereotype.Component;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;


@Component
@Data
@Slf4j
public class FuseProtector implements FuseHandler{
    // ConcurrentHashMap是一个线程安全的哈希表。
    // 在C++中，可以使用 std::unordered_map 或 std::unordered_map 结合互斥量来代替 ConcurrentHashMap。
    private ConcurrentHashMap<String, ServiceState> serviceStateCache = new ConcurrentHashMap<>();
    // 检查当前熔断状态，许可发送请求返回true，拦截返回false
    public boolean fuseHandle(String serviceName){
        ServiceState serviceState = serviceStateCache.get(serviceName);
        switch (serviceState.getFuseState()) {
            case CLOSE:
                return true;
            case FALL_OPEN:
                return false;
            case HALF_OPEN:
                return Math.random() > serviceState.getInterceptRate();
        }
        return true;
    }

    // 初始化服务数据
    public void initCache(ConcurrentHashMap<String, List<ServiceMeta>> serviceList){
        for (String serviceName : serviceList.keySet()) {
            ServiceState serviceState = new ServiceState(serviceName);
                serviceStateCache.put(serviceName,serviceState);
        }
    }

    // 为服务添加一次成功请求次数
    @Async
    @Override
    public void incrSuccess(String serviceName){
        ServiceState serviceState = serviceStateCache.get(serviceName);
        serviceState.incrRequest();
    }

    // 为服务添加一次异常请求次数
    @Override
    public void incrExcept(String serviceName){
        ServiceState serviceState = serviceStateCache.get(serviceName);
        serviceState.incrExcepts();
    }

    // 数据定期刷新，确保全开状态下进行重新检查
    @Scheduled(cron="0/10 * *  * * ? ")
    public void refreshCache(){
        serviceStateCache.replaceAll((n, v) -> new ServiceState(n));
    }
}
